components/collections/src/codepointinvlist/cpinvlist.rs

```diff
@@ -49,14 +49,55 @@ impl<'de: 'a, 'a> serde::Deserialize<'de> for CodePointInversionList<'a> {
         D: serde::Deserializer<'de>,
     {
         use serde::de::Error;
-        let parsed_inv_list = ZeroVec::<u32>::deserialize(deserializer)?;
 
-        CodePointInversionList::try_from_inversion_list(parsed_inv_list).map_err(|e| {
-            Error::custom(format!(
-                "Cannot deserialize invalid inversion list for CodePointInversionList: {:?}",
-                e
-            ))
-        })
+        if deserializer.is_human_readable() {
+            #[derive(serde::Deserialize)]
+            #[serde(untagged)]
+            pub enum De<'data> {
+                #[serde(borrow)]
+                OldStyle(ZeroVec<'data, u32>),
+                #[serde(borrow)]
+                NewStyle(Vec<alloc::borrow::Cow<'data, str>>),
+            }
+
+            match De::<'de>::deserialize(deserializer)? {
+                De::OldStyle(parsed_inv_list) => CodePointInversionList::try_from_inversion_list(parsed_inv_list).map_err(|e| {
+                    Error::custom(format!(
+                        "Cannot deserialize invalid inversion list for CodePointInversionList: {:?}",
+                        e
+                    ))
+                }),
+                De::NewStyle(parsed_strings) => {
+                    let mut ranges = ZeroVec::new_owned(Vec::with_capacity(parsed_strings.len()*2));
+                    let mut size = 0;
+                    for range in parsed_strings {
+                        let mut chars = range.chars();
+                        let (start, end) = match (chars.next(), chars.next(), chars.next(), chars.next()) {
+                            (Some(single), None, None, None) => (single as u32, single as u32 + 1),
+                            (Some(start), Some('-'), Some(end), None) => (start as u32, end as u32 + 1),
+                            _ => return Err(Error::custom(format!(
+                                "Cannot deserialize invalid inversion list for CodePointInversionList: {:?}",
+                                range
+                            )))
+                        };
+                        size += end - start;
+                        ranges.with_mut(|v| {
+                            v.push(start.to_unaligned());
+                            v.push(end.to_unaligned());
+                        });
+                    }
+                    Ok(unsafe {CodePointInversionList::from_parts_unchecked(ranges, size as usize)})
```

Comment:
Suggestion: It would be cleaner to implement this as a conversion from an iterator of ranges, like `FromIterator<(u32, u32)>`. Our FromIterator impls typically perform sorting and other validation.

Reply:
I did it like this because we can compute the size during parsing, but I guess we don't care about JSON performance that much.

Reply:
`FromIterator` is infallible and cannot be implemented on `Result<_, _>`, so this won't work.